src/ReverseProxy/Service/Management/ProxyConfigManager.cs

```
@@ -6,51 +6,122 @@
         }
 
+        // EndpointDataSource
+
+        /// <inheritdoc/>
+        public override IReadOnlyList<Endpoint> Endpoints => Volatile.Read(ref _endpoints);
+
+        /// <inheritdoc/>
+        public override IChangeToken GetChangeToken() => Volatile.Read(ref _changeToken);
+
+        // IProxyConfigManager
+
         /// <inheritdoc/>
-        public async Task ApplyConfigurationsAsync(CancellationToken cancellation)
+        public async Task<EndpointDataSource> LoadAsync()
+        {
+            // Trigger the first load immediately and throw if it fails.
+            // We intend this to crash the app so we don't try listening for further changes.
+            var config = _provider.GetConfig();
+            await ApplyConfigAsync(config);
+
+            if (config.ChangeToken.ActiveChangeCallbacks)
+            {
+                _changeSubscription = config.ChangeToken.RegisterChangeCallback(ReloadConfigAsync, this);
+            }
+
+            return this;
+        }
+
+        // Throws for validation failures
+        private async Task ApplyConfigAsync(IProxyConfig config)
         {
-            var config = await _configBuilder.BuildConfigAsync(cancellation);
+            var dynamicConfig = await _configBuilder.BuildConfigAsync(config.Routes, config.Clusters, default);
 
-            UpdateRuntimeClusters(config);
-            UpdateRuntimeRoutes(config);
+            UpdateRuntimeClusters(dynamicConfig);
+            UpdateRuntimeRoutes(dynamicConfig);
+        }
+
+        private static async void ReloadConfigAsync(object state)
+        {
+            var manager = (ProxyConfigManager)state;
+
+            IProxyConfig newConfig;
+            try
+            {
+                newConfig = manager._provider.GetConfig();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorReloadingConfig(manager._logger, ex);
+                // If we can't load the config then we can't listen for changes anymore.
```

This is catastrophic, shouldn't we let it all explode and crash in this case?

---

That doesn't seem like the kind of choice a library should be making for you. @samsp-msft ?

---

I'd expect the library to either (a) give me chance to recover (i.e. make it possible to react to future updates); or (b) fail fast, not pretending everything is fine. Not doing either of these seems error prone and makes this exceedingly hard to detect in a production environment.

---

A few points:
A) Yes it's bad if the reloads ever stop happening.
B) It should be pretty easy for IProxyConfigProvider implementations to avoid this situation. They're not supposed to signal the prior notification until everything is ready, all GetConfig has to do is return pre-prepped data.
C) We don't have to make a decision about this right now. I'll file an follow-up issue and we can work out how to best handle catastrophic failures.

---

https://github.com/microsoft/reverse-proxy/issues/368